sycl/source/detail/graph_impl.cpp

```diff
@@ -913,6 +912,149 @@ exec_graph_impl::enqueue(const std::shared_ptr<sycl::detail::queue_impl> &Queue,
   return QueueEvent;
 }
 
+void exec_graph_impl::duplicateNodes() {
```

Comment:
This is complex enough that it might be worth having a subgraph section in the design doc about what we're doing at a high level.